### ISSUE TITLE
Cilium unable to remove taints on Azure AKS

### ISSUE THREAD

**Issue by akostic-kostile** [cilium/cilium#19788]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

It seems that Microsoft in their infinite wisdom made a breaking change in 24.04.22. release of AKS regarding taints removal.

https://github.com/Azure/AKS/releases/tag/2022-04-24

`Taints and labels applied using the AKS nodepool API are not modifiable from the Kubernetes API and vice versa. Also, any modifications to system taints will not be allowed.`

Our clusters were deployed using Terraform, which is how we're setting this taint on the nodes (in other words by using nodepool API and not K8S API). 

I found this out 30 minutes ago when, during production deployment, pods were stuck in pending state.

I'm not yet sure how to handle this situation, a workaround that comes to mind might be to provision a new node pool with no taint applied, to manually apply taint using kubectl as Cilium should be able to remove this taint. 

In any case I think this is something you should be aware of and possibly update documentation accordingly.

### Cilium Version

`❯ cilium version
cilium-cli: 0.11.1 compiled with go1.18.1 on darwin/arm64
cilium image (default): v1.11.3
cilium image (stable): v1.11.4
cilium image (running): v1.11.3`

### Kernel Version

`5.4.0-1077-azure`

### Kubernetes Version

`❯ k version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.6", GitCommit:"ad3338546da947756e8a88aa6822e9c11e7eac22", GitTreeState:"clean", BuildDate:"2022-04-14T08:41:58Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.6", GitCommit:"07959215dd83b4ae6317b33c824f845abd578642", GitTreeState:"clean", BuildDate:"2022-03-30T18:28:25Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}`

### Relevant log output

```shell
`Errors:           cilium             cilium-sxrrr    controller mark-k8s-node-as-available is failing since 6s (204x): admission webhook "aks-node-validating-webhook.azmk8s.io" denied the request: (UID: 7ffad4f3-9897-4d9a-a87d-d5af33c87e81) Taint delete request "node.cilium.io/agent-not-ready=true:NoSchedule" refused. User is attempting to delete a taint configured on aks node pool "d16asv5".`
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by atykhyy** [comment]
The [recommended workaround](https://github.com/Azure/AKS/issues/2934#issuecomment-1122837301) is to add a mutating admission webhook which will taint new nodes. I suppose this webhook can run on the system node pool.

**Comment by akostic-kostile** [comment]
@atykhyy Can you give me some more info regarding this? My understanding is, once you register a mutating webhook every request that matches certain conditions, in this case a new node being registered, Kubernetes API will send that information as json to some sort of controller/operator running inside or outside the cluster? In that case we would need either write this controller ourselves or use some 3rd party solution? Is there any 3rd party solution commonly used for this kind of stuff?
Or am I not understanding the situation correctly and everything can be done using just Kubernetes internals?

**Comment by akostic-kostile** [comment]
Problem is, I can't think of a good way to automate this completely. Previously I'd just do `terraform apply` and I'd get fully functional cluster with system + user nodes (these nodes had correct taints set by AKS API). After that it was a matter of running the install script or something like ArgoCD to install everything in the cluster, starting with Cilium.

Now we have this very awkward situation where I have to provision cluster half way, without user nodes. After that comes the installation of kubemod, then we need to continue with Terraform to attach user nodes so those nodes would end up with correct taints set. 

I have to hand it to Microsoft, they sure found a way to de-automate something that was working perfectly fine a week ago.

**Comment by akostic-kostile** [comment]
There is an excellent workaround posted in AKS repo: https://github.com/Azure/AKS/issues/2934#issuecomment-1140613111

Which brings me to question what we should do with this issue? In my opinion at least the documentation should be updated to reflect this new change and then we can close it.

**Comment by crgarcia12** [comment]
> There is an excellent workaround posted in AKS repo: [Azure/AKS#2934 (comment)](https://github.com/Azure/AKS/issues/2934#issuecomment-1140613111)
> 
> Which brings me to question what we should do with this issue? In my opinion at least the documentation should be updated to reflect this new change and then we can close it.

As it is stated few comments bellow, this is going to break in future releases.

**Comment by akostic-kostile** [comment]
True, but it seems they are considering adding "startup taints" (https://github.com/Azure/AKS/issues/2934#issuecomment-1151621093) which should solve this issue in an a better way than an admission controller. It is also a perfect example of over-engineering in my opinion, as the old way was working perfectly fine until they decided to change it for no good reason.

**Comment by sayboras** [comment]
For BYOCNI, there is on going effort as mentioned https://github.com/cilium/cilium/pull/19379.

If you can give it a try, and let us know if it's working. Thanks.

**Comment by davem-git** [comment]
> 1. https://raw.githubusercontent.com/kubemod/kubemod/v0.15.0/bundle.yaml

what removes it when cilium is stood up properly?

**Comment by davem-git** [comment]
@akostic-kostile https://github.com/cilium/cilium/issues/19788#issuecomment-1125110915. I'm attempting to go down this path. can't seem to get  kubemod to fire. no logs that i can see. Did you make any adjustments to your setup? 

```sh
{"level":"info","ts":"2022-10-11 22:35:40.391Z","logger":"webapp-setup","msg":"web app server is starting to listen","addr":":8081"}
{"level":"info","ts":"2022-10-11 22:35:40.695Z","logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":":8082"}
{"level":"info","ts":"2022-10-11 22:35:40.695Z","logger":"operator-setup","msg":"health server is starting to listen","addr":":8083"}
{"level":"info","ts":"2022-10-11 22:35:40.696Z","logger":"controller-runtime.builder","msg":"Registering a mutating webhook","GVK":"api.kubemod.io/v1beta1, Kind=ModRule","path":"/mutate-api-kubemod-io-v1beta1-modrule"}
{"level":"info","ts":"2022-10-11 22:35:40.696Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/mutate-api-kubemod-io-v1beta1-modrule"}
{"level":"info","ts":"2022-10-11 22:35:40.696Z","logger":"controller-runtime.builder","msg":"Registering a validating webhook","GVK":"api.kubemod.io/v1beta1, Kind=ModRule","path":"/validate-api-kubemod-io-v1beta1-modrule"}
{"level":"info","ts":"2022-10-11 22:35:40.696Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/validate-api-kubemod-io-v1beta1-modrule"}
{"level":"info","ts":"2022-10-11 22:35:40.696Z","logger":"operator-setup","msg":"registering core mutating webhook"}
{"level":"info","ts":"2022-10-11 22:35:40.696Z","logger":"controller-runtime.webhook","msg":"registering webhook","path":"/dragnet-webhook"}
{"level":"info","ts":"2022-10-11 22:35:40.696Z","logger":"operator-setup","msg":"starting manager"}
{"level":"info","ts":"2022-10-11 22:35:40.697Z","logger":"controller-runtime.manager","msg":"starting metrics server","path":"/metrics"}
{"level":"info","ts":"2022-10-11 22:35:40.697Z","logger":"controller","msg":"Starting EventSource","reconcilerGroup":"api.kubemod.io","reconcilerKind":"ModRule","controller":"modrule","source":"kind source: /, Kind="}
{"level":"info","ts":"2022-10-11 22:35:40.697Z","logger":"controller-runtime.webhook.webhooks","msg":"starting webhook server"}
{"level":"info","ts":"2022-10-11 22:35:40.698Z","logger":"controller-runtime.certwatcher","msg":"Updated current TLS certificate"}
{"level":"info","ts":"2022-10-11 22:35:40.789Z","logger":"controller-runtime.webhook","msg":"serving webhook server","host":"","port":9443}
{"level":"info","ts":"2022-10-11 22:35:40.790Z","logger":"controller-runtime.certwatcher","msg":"Starting certificate watcher"}
{"level":"info","ts":"2022-10-11 22:35:40.897Z","logger":"controller","msg":"Starting Controller","reconcilerGroup":"api.kubemod.io","reconcilerKind":"ModRule","controller":"modrule"}
{"level":"info","ts":"2022-10-11 22:35:40.897Z","logger":"controller","msg":"Starting workers","reconcilerGroup":"api.kubemod.io","reconcilerKind":"ModRule","controller":"modrule","worker count":1
```

**Comment by adwittumuluri** [comment]
v1.14 of [`the cilium-operator-azure`](https://docs.cilium.io/en/latest/cmdref/cilium-operator-azure/) supports a `--pod-restart-selector` argument that, when empty, tells the Cilium operator to restart all unmanaged pods. We were able to leverage this in the absence of taints, and it seems to be fine, albeit not as preferred as the `node.cilium.io/agent-not-ready=true:NoSchedule` option.

In addition to this, we created and install our own `node-mutating-webhook` to inject the `node.cilium.io/agent-not-ready:NoExecute` taint upon node creation. This mirrors the desired behavior on other clouds for installing Cilium.

Both approaches can happily run simultaneously, so we're keeping both around in case AKS decides to pull the rug out from underneath us again.
